src/libraries/Microsoft.PowerFx.Interpreter/UDFTexlFunction.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx
+{
+    public sealed partial class RecalcEngine
+    {
+        private class UDFTexlFunction : CustomTexlFunction, IAsyncTexlFunction
+        {
+            internal NamedFormulaType[] _parameterNames;
+            internal IExpression _expr;
+            internal LazyCheck _check;
+
+            public UDFTexlFunction(string name, FormulaType returnType, params FormulaType[] paramTypes)
+                : base(name, returnType, paramTypes)
+            {
+            }
+
+            public async Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancel)
+            {
+                // $$$ There's a lot of unnecessary string packing overhead here 
+                // because Eval wants a Record rather than a resolved arg array.                 
+                var parameters = FormulaValue.NewRecordFromFields(Zip(_parameterNames, args));
+
+                var result = await LazyGetExpression().EvalAsync(parameters, cancel);
+
+                return result;
+            }
+
+            public IExpression LazyGetExpression()
+            {
+                if (_expr == null)
+                {
+                    var check = _check.Get();
+                    check.ThrowOnErrors();
+                    _expr = check.Expression;
+                }
+
+                return _expr;
+            }
+        }
+
+        private class LazyCheck
+        {
+            internal string _expressionText;
+            internal RecordType _parameterType;
+            internal ParserOptions _options;
+            internal Engine _engine;
+
+            public LazyCheck(Engine engine, string expressionText, RecordType parameterType = null, ParserOptions options = null)
+            {
+                _engine = engine;
+                _expressionText = expressionText;
+                _parameterType = parameterType;
+                _options = options;
+            }
+
+            public CheckResult Get() => _engine.Check(_expressionText, _parameterType, _options);
+        }
+
+        internal void DefineFunction(string name, string body, FormulaType returnType, params NamedFormulaType[] parameters)
```

Comment:
>DefineFunction [](http://example.com/codeflow?start=22&length=14)

This should probably be in recalcengine.cs

Reply:
Fixed